### DISCUSSION TITLE
What are the advantages of using this library over React?

### DISCUSSION THREAD

**Discussion by larsbs** [solidjs/solid#582] [Q&A]
Hi, I've been looking for alternatives to React for a while and skimming the README of the library, it looks really promising. However, I don't fully see what are the advantages/benefits of using it over React. I checked the "[comparison with other libraries](https://github.com/solidjs/solid/blob/main/documentation/comparison.md)" section but it just goes over how the library is different to React in technical terms, nothing about DX, differences in reasoning or things that are complicated in React but this library makes easier.

So in summary and putting it bluntly, why should I be using this library instead of React? Is there a possibility of expanding into this? and I think a section like that would also help with the marketability of the library as well 😄

**Comment by AjaxSolutions** [comment]
Solid has more granular updates, so performance will be better.

https://dev.to/ryansolid/5-ways-solidjs-differs-from-other-js-frameworks-1g63

_"This is by far the most defining part of Solid. Components do not re-run, just the primitives and JSX expressions you use. This means no stale closures or Hook Rules for those of you coming from React."_

**Comment by excessivecoding** [comment]
I'm personally in the same boat. While I understand the technical changes between React and Solid with all the benefits. I would love to know what would be the philosophy behind the framework in the future.

Is Solid mean to stay the smallest possible and unopinionated à la React or is the plan to create an official and open ecosystem like Vue.js (Vuex, Vue-Router, etc.)

Will Solid optimize on Developer Experience like Svelte (e.g. reactive sugar syntax) while trading-off something else in return ?

It would be great to have some insight besides technical and performances choices that would help people get onboard with Solid.js in the future.

**Comment by ryansolid** [comment]
Yeah that's a good point. The technical ones are easy because it's something that React can't do nor really follow without a lot of breaking changes. So that alone is something that makes Solid's approach standout. And even the DX advantages translate almost immediately as you move away from say Hook rules.

The problem is the rest is a bit more open to opinion. I've written at length in articles about design decisions but that is too much for someone to just look at quickly. I can think about how I can summarize that. I do have a section in getting started that talks about the philosophy around Solid: https://www.solidjs.com/docs/1.0.0#think-solid. But I don't go as far as calling React out. I've gotten a lot of complaints in the past about how React slanted the docs are, and I've tried to remedy that. That being said having a special place or at least an article might help someone recognize the differences without actually needing to try the library. It is a bit difficult to explain to anyone who doesn't have the React mental model imprinted on their brain. I think I need to come up with concrete examples. I'm working on it as this is the first thing that comes up on almost all streams I do. So I'm getting better at telling that story but it not only takes practice, but experimentation to see what resonates. Because trust me I've explained this 1000 different ways at this point and I still don't feel I have it nailed down.

**Reply by brillout** [reply]
> I've gotten a lot of complaints in the past about how React slanted the docs are, and I've tried to remedy that.

Since both Solid and React use JSX, it makes a lot of sense to directly compare both. While giving credits to React where it's due (e.g. I believe React's RSC is superior to what I've seen so far on Solid's side).

**Comment by windmaomao** [comment]
This is my question while playing with Solid while thinking about React. 

React relies on the fiber tree heavily therefore one of the feature is that a fiber node can be interrupted if a high priority request comes to the picture. 

With Solid, I guess there's no fiber (apology haven't read the source code yet), only DOM. So how can a "render" in solid be interrupted? or maybe the question is that do we need to care about the concurrent picture that React painted for release 18. 

I wonder maybe we need a demo to do a heavy calculation while typing in the keyboard to keep interrupt it. I believe i saw this type of example somewhere, and see how Solid performs. Maybe we will run into that issue as well, but in that case, how should Solid address this problem? 

Sorry, i'm not trying to be picky here, just wondering if React heads to the wrong direction :)

**Reply by ryansolid** [reply]
Fiber in React lets them break apart rendering. Interupt (and restart work). But it's still at a certain granularity. With Solid everything is already granular. Like fiber on the outside. One simply delays a node updating. You've probably seen this demo before.. where each circle adds a delay. https://ryansolid.github.io/solid-sierpinski-triangle-demo/ . I've implemented it in Solid with the same delay and using our `createDeferred` scheduler.

The thing is granular updates get you so much for free before you hit that point. You've probably seen this demo before too: https://twitter.com/0xca0a/status/1199997552466288641.

Solid has the most performant version with no scheduling tricks: https://codesandbox.io/s/solid-gl-boxes-lmpei?file=/index.jsx

I'm still working on a cool demo to show off something similar with scheduling. It's just harder to make it visual because Solid already is very performant. But there is a miniature version of React's fiber scheduler in Solid that can be used through `requestCallback` and `createDeferred`.

**Reply by windmaomao** [reply]
thank you very much @ryansolid , this is exactly what i'm looking for. The part i was trying to understand is the interrupt of fiber update. 

"With Solid everything is already granular." This is  the part I need to study a bit further.  For example, there's another example i was having in mind. If there's a big render I need to make, and the keyboard is keep interrupting the update, how does Solid perform. 

I took a look at the triangle demo (performance tab), seems have very good performance, can i ask where is the source code? thank you.

**Reply by ryansolid** [reply]
Sorry I linked the wrong thing. Here: https://github.com/ryansolid/solid-sierpinski-triangle-demo.

I sort went off on this subject on Twitter and our discord yesterday because I had been in search of demos to really showcase it and was having a hard time. If you are on Twitter this thread is great. Both Dan Abramov and Paul Henschel weighed in: https://twitter.com/RyanCarniato/status/1422479172836855812

**Comment by atk** [comment]
The developer experience in react is heavily reliant on tooling (react dev tools). At the moment, there are no dev tools for solid, but on the other hand, the transpiled components are really readable and breakpoints don't trigger every other second, which eases debugging. Still, the point goes to react here, because the tooling is quite superb.

On the other hand, some of react's limitations, e.g. the inability to conditionally include a hook (there is a workaround, but it's not nice), don't exist in solid. There are also no render cycles to be interrupted, since solid does not walk a render tree all the time but instead lets its observables call the actions required to render the required changes - and if you don't want to run anything, you can use the equals callback to stop the observable from triggering (it could also block updates depending on external factors). This and the lot of reactive tools of solid lead to very controllable and predictable behavior. And while you have to rely on refs to set properties like srcObject in react, you can just use prop: in solid. In my opinion, solid gets the point for this one.

The icing of every framework's cake is ecosystem and community. The sheer amount of libraries, tools, helpers, components available for react is almost too large to navigate. Here, solid is a bit lacking, but quietly catching up. The most important tools are already available. On the other hand, while the react community with people like Kent C. Dodds is very positive and welcoming, it is rather difficult to contribute to react itself. I found the community very open and helpful and the creator @ryansolid is even dedicating a lot of his valuable time to answer beginner's questions like mine. So while it is a close call, I would award this point to solid.

On the other hand, if you merely need to learn a framework that gets you the most job offers, you should choose react. But I don't think I should award points for "most managers have heard and have a lot of misconceptions about it".

**Reply by davedbase** [reply]
This was a very fair and balanced description of React and Solid. Thanks for weighing in so eloquently!

**Comment by ryansolid** [comment]
I converted this to a discussion to make it easier to respond to people directly. I think this will promote a better dialog and make it easier to extract the keys parts to pull into an article or guide.
